audioipc/src/ipccore.rs

```diff
@@ -0,0 +1,717 @@
+                    assert_eq!(fd.len(), 1);
+                    assert!(handle.is_none());
+                    handle = Some(fd[0]);
+                }
+                item.set_owned_handle(|| handle);
+            }
+            self.handler.consume(item)?;
+        }
+
+        Ok(false)
+    }
+
+    // Caller will try to write `outbound` to associated connection, queuing any data that can't be transmitted immediately.
+    fn flush_outbound(&mut self, outbound: &mut sys::ConnectionBuffer) -> Result<()> {
+        debug!("flush_outbound: {:?}", outbound.buf);
+
+        // Repeatedly grab outgoing items from the handler, passing each to `encode` for serialization into `outbound`.
+        while let Some(mut item) = self.handler.produce()? {
+            let handle = item.take_handle_for_send();
+
+            // On Windows, the handle is transferred by duplicating it into the target remote process during message send.
+            #[cfg(windows)]
+            if let Some((handle, target_pid)) = handle {
+                let remote_handle = unsafe { duplicate_platform_handle(handle, Some(target_pid))? };
+                trace!(
+                    "item handle: {:?} remote_handle: {:?}",
+                    handle,
+                    remote_handle
+                );
+                // The new handle in the remote process is indicated by updating the handle stored in the item with the expected
+                // value on the remote.
+                item.set_remote_handle_value(|| Some(remote_handle));
+            }
+            // On Unix, the handle is encoded into a cmsg buffer for out-of-band transport via sendmsg.
+            #[cfg(unix)]
+            if let Some((handle, _)) = handle {
+                item.set_remote_handle_value(|| Some(handle));
+            }
+
+            self.codec.encode(item, &mut outbound.buf)?;
+
+            #[cfg(unix)]
+            if let Some((handle, _)) = handle {
+                // TODO: Rework builder to commit directly to outbound buffer.
+                match cmsg::builder(&mut outbound.cmsg).rights(&[handle]).finish() {
+                    Ok(handle_bytes) => outbound.cmsg.extend_from_slice(&handle_bytes),
+                    Err(e) => debug!("cmsg::builder failed: {:?}", e),
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+struct FramedDriver<T: Handler> {
+    codec: LengthDelimitedCodec<T::Out, T::In>,
+    handler: T,
+}
+
+impl<T: Handler> FramedDriver<T> {
+    fn new(handler: T) -> FramedDriver<T> {
+        FramedDriver {
+            codec: Default::default(),
+            handler,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EventLoopThread {
+    thread: Option<thread::JoinHandle<Result<()>>>,
+    handle: EventLoopHandle,
+}
+
+// TODO: Builder pattern.
+impl EventLoopThread {
+    pub fn new<F1, F2>(
+        name: String,
+        stack_size: Option<usize>,
+        after_start: F1,
+        before_stop: F2,
```

Comment:
All existing callers supply `F1` and `F2`.  Making them optional clutters the call site since you need to explicitly specify the type of `F1` and `F2` in the `None` case - that could be solved with a builder pattern, but since they're currently always specified it doesn't seem worth the added complexity.